### PR TITLE
[BEAM-13695] Add option to add-open flags for Java container

### PR DESCRIPTION
**Pull request by kileys**
Changes:
Add JDK add-open flag to pipeline options
Add ability to pass in json files to update java args, properties and classpath
Add jamm java args to boot script from json file

Tested:
Ran integration tests that uses boot script for Java 8/17
Docker logs from https://ci-beam.apache.org/job/beam_LoadTests_Java_CoGBK_Dataflow_V2_Batch_Java17_PR/6 includes the new jamm options in the execution
https://pantheon.corp.google.com/logs/query;query=resource.type%3D%22dataflow_step%22%0Aresource.labels.job_id%3D%222022-03-22_14_23_07-6960246593216693989%22%0Alog_name%3D%22projects%2Fapache-beam-testing%2Flogs%2Fdataflow.googleapis.com%252Fdocker%22%0A%22jamm%22;timeRange=2022-03-22T21:23:07.646Z%2F2022-03-22T21:29:54.883Z;cursorTimestamp=2022-03-22T21:25:05.743895120Z?project=apache-beam-testing
[WIP] Add test that uses new pipeline option (manually tested locally)

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by kileys** [comment]
Run Load Tests Java 17 CoGBK Dataflow V2 Batch

**Comment by kileys** [comment]
Run Load Tests Java 17 CoGBK Dataflow V2 Batch

**Comment by kileys** [comment]
Run Load Tests Java 17 CoGBK Dataflow V2 Batch

**Comment by kileys** [comment]
R: @lukecwik 
CC: @kennknowles

**Comment by kileys** [comment]
Checked that there's no more warnings from the cache. I'll look to add an integration test for it

**Comment by kileys** [comment]
Run Load Tests Java 17 CoGBK Dataflow V2 Batch

**Comment by kileys** [comment]
Run Load Tests Java 11 CoGBK Dataflow V2 Batch

**Comment by kileys** [comment]
Run Load Tests Java 17 CoGBK Dataflow V2 Batch

**Comment by lukecwik** [comment]
R: @damccorm @jrmccluskey

Can either of you review the Go code?

I'll review the Java changes.

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptions.java [hunk 1]
```diff
@@ -217,4 +217,22 @@ public String create(PipelineOptions options) {
   boolean isHotKeyLoggingEnabled();
 
   void setHotKeyLoggingEnabled(boolean value);
+
+  /**
+   * Open modules needed for reflection that access JDK internals with Java 17+
```
Comment (lukecwik) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/options/DataflowPipelineOptions.java:222]:
```suggestion
   * Open modules needed for reflection that access JDK internals with Java 9+
```

File: sdks/java/core/src/main/java/org/apache/beam/sdk/options/SdkHarnessOptions.java [hunk 2]
```diff
@@ -294,4 +295,18 @@ public static SdkHarnessLogLevelOverrides from(Map<String, String> values) {
       return overrides;
     }
   }
+
+  /**
+   * Open modules needed for reflection that access JDK internals with Java 17+
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/options/SdkHarnessOptions.java:300]:
```suggestion
   * Open modules needed for reflection that access JDK internals with Java 9+
```

File: sdks/java/core/src/main/java/org/apache/beam/sdk/options/SdkHarnessOptions.java [hunk 3]
```diff
@@ -294,4 +295,18 @@ public static SdkHarnessLogLevelOverrides from(Map<String, String> values) {
...
+   *
+   * <p>With JDK 16+, <a href="#{https://openjdk.java.net/jeps/403}">JDK internals are strongly
+   * encapsulated</a> and can result in an InaccessibleObjectException being thrown if a tool or
+   * library uses reflection that access JDK internals. If you see these errors in your worker logs,
+   * you can pass in modules to open using the format module/package=target-module(,target-module)*
+   * to allow access to the library. E.g. java.base/java.lang=jamm
+   */
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/options/SdkHarnessOptions.java:307]:
```suggestion
   *
   * <p>You may see warnings that jamm, a library used to more accurately size objects, is unable to
   * make a private field accessible. To resolve the warning, open the specified module/package to
   * jamm.
   */
```

File: sdks/java/container/boot.go [hunk 4]
```diff
@@ -211,3 +241,120 @@ func heapSizeLimit(info *fnpb.ProvisionInfo) uint64 {
...
+		return nil
+	}
+
+	if err := filepath.Walk(dir, worker); err != nil {
+		return nil, err
+	}
+	return meta, nil
+}
+
+func BuildOptions(metaOptions []*MetaOption) *Options {
```
Comment (damccorm) [inline sdks/java/container/boot.go:337]: Could you please add tests for this and LoadMetaOptions? Both are pretty well contained/good candidates for units.

File: sdks/java/container/boot.go [hunk 5]
```diff
@@ -193,7 +195,35 @@ func main() {
 	} else {
 		args = append(args, jammAgentArgs)
 	}
+	// Apply meta options
+	const metaDir = "/opt/apache/beam/options"
+	metaOptions, err := LoadMetaOptions(metaDir)
+	javaOptions := BuildOptions(metaOptions)
+	// (1) Add custom jvm arguments: "-server -Xmx1324 -XXfoo .."
+	args = append(args, javaOptions.JavaArguments...)
```
Comment (damccorm) [inline sdks/java/container/boot.go:203]: Is there any risk of duplicate args here? What would happen in that case/should we do some validation/warn or overwrite?

**Review by damccorm — COMMENTED** [review]
This mostly looks good to me (sans a couple comments). This PR would benefit from a high level description of the change and a description of how it was tested though - since it crosses area boundaries its pretty hard to tell if it will actually work through code review alone, plus it helps optimize for reviewers and future readers (I know at least I would have benefited from description of changes/testing done 😃).

File: sdks/java/container/boot.go (hunk 5 above)
Comment (kileys) [reply sdks/java/container/boot.go:203]: There may be duplicate args, but there's a priority associated with the MetaOptions, so higher priority will override the lower

**Comment by kileys** [comment]
Thanks for reviewing!

@damccorm updated the pr comment with changes/testing

**Comment by kileys** [comment]
Run Load Tests Java 17 CoGBK Dataflow V2 Batch

**Comment by kileys** [comment]
Dataflow Runner V2 Java 17 Nexmark Tests

**Comment by kileys** [comment]
Run Dataflow Runner V2 Java 17 Nexmark Tests

**Comment by kileys** [comment]
Run Dataflow Runner V2 Java 17 Nexmark Tests

**Comment by lukecwik** [comment]
Run Java PreCommit

File: sdks/java/container/common.gradle [hunk 6]
```diff
@@ -74,6 +74,13 @@ task copyGolangLicenses(type: Copy) {
     dependsOn ':release:go-licenses:java:createLicenses'
 }
 
+task copyJdkOptions(type: Copy) {
+    if (imageJavaVersion == "17") {
```
Comment (lukecwik) [inline sdks/java/container/common.gradle:78]: As a follow up we should do this for jdk 11 as well.

File: sdks/java/container/common.gradle [hunk 7]
```diff
@@ -108,3 +115,4 @@ if (project.rootProject.hasProperty(["docker-pull-licenses"]) ||
 dockerPrepare.dependsOn copySdkHarnessLauncher
 dockerPrepare.dependsOn copyDockerfileDependencies
 dockerPrepare.dependsOn ":sdks:java:container:downloadCloudProfilerAgent"
+dockerPrepare.dependsOn copyJdkOptions
```
Comment (lukecwik) [inline sdks/java/container/common.gradle:118]: nit New line at eof

**Comment by lukecwik** [comment]
https://ci-beam.apache.org/job/beam_PostCommit_Java_Nexmark_DataflowV2_Java17_PR/7/ and https://ci-beam.apache.org/job/beam_LoadTests_Java_CoGBK_Dataflow_V2_Batch_Java17_PR/6/ passed. GitHub UI failed to update based upon the Jenkins job status.

**Comment by lukecwik** [comment]
Kiley, please follow up with Java 11 change, nits, and tests.

File: sdks/java/container/boot.go (hunk 4 above)
Comment (damccorm) [reply sdks/java/container/boot.go:337]: Yeah, you should be able to just add a `boot_test.go` in the same directory with `package main`, then you can add functions like `Test<function you are testing>(t *testing.T)` which will get automatically picked up by go when you run go test in the directory. Here's an example test file - https://github.com/apache/beam/blob/master/sdks/go/pkg/beam/core/funcx/fn_test.go

**Pull request merged**